### PR TITLE
ARROW-8651: [Python][Dataset] Support pickling of Dataset objects

### PR DESCRIPTION
**Pull request by kszucs**

### CODE REVIEW

File: python/pyarrow/_dataset.pyx [hunk 1]
```diff
@@ -773,6 +789,14 @@ cdef class FileFragment(Fragment):
         Fragment.init(self, sp)
         self.file_fragment = <CFileFragment*> sp.get()
 
+    def __reduce__(self):
+        buffer = self.buffer
+        return self.format.make_fragment, (
```
Comment (jorisvandenbossche) [inline python/pyarrow/_dataset.pyx:794]: By specifying here the method on a format object (`format.make_fragment`), it also automatically pickles the `format` *instance* ?

File: python/pyarrow/_dataset.pyx [hunk 2]
```diff
@@ -887,6 +911,14 @@ cdef class ParquetFileFragment(FileFragment):
         FileFragment.init(self, sp)
         self.parquet_file_fragment = <CParquetFileFragment*> sp.get()
 
+    def __reduce__(self):
+        return self.format.make_fragment, (
+            self.path,
```
Comment (jorisvandenbossche) [inline python/pyarrow/_dataset.pyx:916]: I suppose you might need to do the same `self.path if buffer is None else buffer,` here as you did for `FileFragment` ?

File: python/pyarrow/_dataset.pyx (hunk 1 above)
Comment (jorisvandenbossche) [reply python/pyarrow/_dataset.pyx:794]: We should maybe ensure this with testing a picking roundtrip for a case that specified read params in the ParquetFileFormat object

**Review by jorisvandenbossche — COMMENTED** [review]
Looks good!

I am fine with leaving properly pickling RowGroupInfo objects (including their metadata) for a follow-up JIRA/PR.

File: python/pyarrow/_dataset.pyx (hunk 2 above)
Comment (kszucs) [reply python/pyarrow/_dataset.pyx]: Fixed it, also added a test.

File: python/pyarrow/_dataset.pyx (hunk 1 above)
Comment (kszucs) [reply python/pyarrow/_dataset.pyx:794]: Adding a case.

Comment (kszucs) [reply python/pyarrow/_dataset.pyx:794]: > By specifying here the method on a format object (`format.make_fragment`), it also automatically pickles the `format` _instance_ ?

Apparently.

File: python/pyarrow/tests/test_dataset.py [hunk 3]
```diff
@@ -635,6 +635,37 @@ def test_make_fragment_from_buffer():
...
+    pq.write_table(table, out)
+
+    buffer = out.getvalue()
+
+    formats = [
+        ds.ParquetFileFormat(),
+        ds.ParquetFileFormat(
+            read_options=ds.ParquetReadOptions(
+                use_buffered_stream=True,
+                buffer_size=4096,
```
Comment (jorisvandenbossche) [inline python/pyarrow/tests/test_dataset.py:657]: we probably need to use an option that actually alters the output to be able to catch a failure, eg `dictionary_columns`

**Comment by jorisvandenbossche** [comment]
I opened https://issues.apache.org/jira/browse/ARROW-9332 for the parquet statistics

File: python/pyarrow/tests/test_dataset.py (hunk 3 above)
Comment (kszucs) [reply python/pyarrow/tests/test_dataset.py]: @jorisvandenbossche updated.

File: python/pyarrow/tests/test_dataset.py [hunk 4]
```diff
@@ -612,6 +613,83 @@ def test_make_fragment(multisourcefs):
...
+def test_make_parquet_fragment_from_buffer():
+    import pyarrow.parquet as pq
+
+    cases = [
+        (
+            pa.table(
+                [
+                    ['a', 'b', 'c'],
+                    [12, 11, 10],
+                    ['dog', 'cat', 'rabbit']
```
Comment (jorisvandenbossche) [inline python/pyarrow/tests/test_dataset.py:648]: you could reduce the used vertical space here a bit by only defining the list of arrays here, and do `table = pa.table(arrays, names=['alpha', 'num', 'animal'])` in the loop

**Review by jorisvandenbossche — APPROVED** [review]

File: python/pyarrow/tests/test_dataset.py (hunk 4 above)
Comment (kszucs) [reply python/pyarrow/tests/test_dataset.py]: Done.

**Pull request closed without merging**